### DISCUSSION TITLE
Issue with loading content ("Offline version" of emulator)

### DISCUSSION THREAD

**Discussion by Cattn** [EmulatorJS/EmulatorJS#448] [Q&A] [answered]
Hello, me and some friends have been designing an offline system with, emulatorjs in mind, and so as part of it, we convert the binary file to base64 and save that into a js file to the user's machine so that it can pulled without having to ask for file permission each time.

So right now, we reference a js file with emulatorjs params, a js file with the encoded base64, and then the githack loaderjs script (in that order in the html)

The emulator loads just fine, but always says "Failed to Load Content" when trying to pull the data after conversion

let coder = new TextEncoder();
            bar = coder.encode(decodeURIComponent(escape(window.atob(content))));
            console.log(bar);
            let bin = new Uint8Array(bar);
            window.gameData = bin; 
            
(variable 'content' is the base64)
Now I think it's the fault of how we decode it, not encoding, nor the emulator. I was wondering if anyone could offer a solution to the issue. I've been at it for a while trying many different ways to decode it into the proper format, and I'm under the impression that window.gameData has to be a Unit8Array. 

EDIT: added more references to files
[Here's the Full JS file/repo](https://github.com/Math-Study/bunker-v1/blob/main/public/js/convert.js)
[Here's the full html file referenced in](https://github.com/Math-Study/bunker-v1/blob/main/public/folders/emulators/emulatorjs.html)
[Here's the rom referenced](https://github.com/Math-Study/bunker-v1/blob/main/public/my-lib/advancewars.gba)
[Here's the JS file that the rom was encoded and stored in](https://github.com/Math-Study/bunker-v1/blob/main/public/my-lib/advancewars.js)
[Here's the encoder](https://github.com/Math-Study/bunker-v1/blob/main/public/folders/lib/index.html)

**Comment by ethanaobrien** [comment, marked as answer]
What is the type of the `content` variable? Is it a Uint8Array? ArrayBuffer? Blob?

**Reply by ethanaobrien** [reply]
Sorry, just saw that you said `content` is a base 64. Try this
```
JSON.stringify(Array.from(new Uint8Array(await (await fetch('data:text/plain;base64,'+content)).arrayBuffer())));
```

**Reply by Cattn** [reply]
Alright Will try, had internet issues most the day

**Reply by Cattn** [reply]
`Refused to apply style from 'http://127.0.0.1:5500/public/folders/emulators/emu-css.min.css?v=30.5' because its MIME type ('text/html') is not a supported stylesheet MIME type, and strict MIME checking is enabled.`

I know this error is caused due to an issue with my code, sorry I'm still tryna learn how async/await works so I'm not fully knowledgeable on how to piece it together

window.globalGameIntercept = {};
        globalGameIntercept.register = function(name, content){
        convert(content);         
  }

       
  async function convert(content) {
        let bin = JSON.stringify(Array.from(new Uint8Array(await (await fetch('data:text/plain;base64,'+content)).arrayBuffer())));
        window.gameData = bin;
        EJS_player = '#game';
        EJS_core = "gba";
        EJS_DEBUG_XX = true;
        EJS_gameName = 'advancewars';
        EJS_color = "#de4587";
        EJS_pathtodata = "https://rawcdn.githack.com/EmulatorJS/EmulatorJS/main/data/";
        EJS_gameUrl = URL.createObjectURL(new Blob([Uint8Array.from(window.gameData)]));
        console.log(bin);
}

**Reply by Nailington** [reply]
Unrelatedish... (didn't see @Cattn send another message, my bad) I am also working on the project with him (I know you didn't see that one coming 😄). I am unable to use the rawcdn/githack link on the internet connection at our school. Would we be able/allowed to use download it and host it ourselves so we can use it?

**Reply by ethanaobrien** [reply]
>Would we be able/allowed to use download it and host it ourselves so we can use it?

As long as you state the source (which it does in the context menu) then yes!

Just wondering, is the official https://emulatorjs.org website blocked?


@Cattn The url is being set properly, its not your code. It actually seems to be a problem with the web server you are using. What server are you using to host this?

**Reply by Cattn** [reply]
I'm using vscode live server for testing, should I try through file://?

**Reply by Cattn** [reply]
I tried it, got:
Failed to load resource: net::ERR_FILE_NOT_FOUND emu-css.min.css?v=30.5

**Reply by Nailington** [reply]
> > Would we be able/allowed to use download it and host it ourselves so we can use it?
> 
> As long as you state the source (which it does in the context menu) then yes!
> 
> Just wondering, is the official https://emulatorjs.org website blocked?

https://emulatorjs.org is NOT blocked. I can access that like normal. And cool to know! I will probably add the rawcdn data to the same repo that is in question right now.

**Reply by ethanaobrien** [reply]
Let me know if adding the code to the repo fixes it please

**Reply by Nailington** [reply]
I am going to do so now! I assume that my issue and @Cattn's issue are unrelated (and there is a high chance it wont matter for now until he fixes his issue). I will let you know though!

**Reply by Cattn** [reply]
Gonna mark this as complete, as I've resolved my issues
